container/exec: Support ConsoleSize

vvoland opened the pull request:
Relates to 
- https://github.com/docker/cli/issues/3554

Now client have the possibility to set the console size of the executed process immediately at the creation. This makes a difference for example when executing commands that output some kind of text user interface which is bounded by the console dimensions.

**- What I did**
Make exec accept the console size and set it for the created exec

**- How I did it**
Introduced ConsoleSize parameter which is copied to specs.Process that is passed to containerd.Exec

**- How to verify it**
- Run the TestExecConsoleSize integration test.
- With CLI changes you can also check with `docker exec`:
```
# resize terminal from the default size

cid=$(docker run -d ubuntu sleep infinity)

docker exec -it $cid stty size
# check if the printed size is as expected

docker exec -it $cid top -n 1
# check if top UI is rendered at the full terminal size
```

**- Description for the changelog**
Exec allows to set the console size for the exec'd process immediately when it's created


**- A picture of a cute animal (not mandatory but encouraged)**
![dudus2](https://user-images.githubusercontent.com/5046555/169486224-a3b12c1c-0878-4ca2-afde-535c7a22a202.jpg)

thaJeztah commented on api/types/configs.go:
```diff
@@ -33,6 +33,7 @@ type ExecConfig struct {
 	User         string   // User that will run the command
 	Privileged   bool     // Is the container in privileged mode
 	Tty          bool     // Attach standard streams to a tty.
+	ConsoleSize  *[2]uint // Initial console size [height, width]
```
I've been going back-and-forth on whether or not we should use a pointer for this one, and could use some input on that bit;


- 👍 we can reset it to `nil` (nice), and we could (probably should) add a `json:.omitempty` to make the API requests a bit cleaner (not sending properties that are not used / default)
- 👉  for `HostConfig.ConsoleSize` we don't use a pointer. This is was the main reason to start this train of thought; do we prefer to be consistent, or do we consider "let's use the better option for this new one?"
- ❓  pointers can be a bit awkward to use, but in this case for a struct it's not really an issue (I did recently remove some `*int` from some structs, as those definitely _are_ a bit of a pain to work with)


So, I _think_ this should be fine (minus us not being consistent), but thought to at least write it down (also for others reviewing)

As to the "omit empty"; I think that'd be a nice thing to add here; WDYT?

```suggestion
	ConsoleSize  *[2]uint `json:",omitempty"` // Initial console size [height, width]
```

vvoland replied:
The omitempty should definitely be there, thanks!
As for the pointer usage, I think it's better to use pointer here, because it makes it clearer that this is optional, which IMO is important, given that this is an API.

thaJeztah replied:
>  I think it's better to use pointer here, because it makes it clearer that this is optional, which IMO is important, given that this is an API.

Yes, agreed. It's just unfortunate we didn't do the same for the other one (possibly something to look into, but not sure if we can change now).

thaJeztah commented on client/container_exec.go:
```diff
@@ -14,6 +14,9 @@ func (cli *Client) ContainerExecCreate(ctx context.Context, container string, co
 	if err := cli.NewVersionError("1.25", "env"); len(config.Env) != 0 && err != nil {
 		return response, err
 	}
+	if err := cli.NewVersionError("1.42", "console size"); config.ConsoleSize != nil && err != nil {
+		return response, err
+	}
```
Perhaps for this one we shouldn't produce an error (we are a bit inconsistent on this though!); my train of thought here is that; unlike `env` above, the console size is not an explicit option set by the user; this option is a bit similar to how we moved the "auto remove"  (`--rm`) from the client to the daemon; https://github.com/moby/moby/blob/b6dab55339382e2b6df0678a511a75c146daa45b/client/container_create.go#L32-L35

Having this produce an error means that we also need to check the API version in `docker/cli`. Perhaps instead we could just set it unconditionally in the CLI (and have the client discard the value), which could be slightly easier to use.

Happy to hear your (and other's) thoughts on that as well though!

thaJeztah commented on daemon/exec.go:
```diff
@@ -237,6 +238,12 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, stdin
 	p.Env = ec.Env
 	p.Cwd = ec.WorkingDir
 	p.Terminal = ec.Tty
+	if p.Terminal && ec.ConsoleSize != nil {
+		p.ConsoleSize = &specs.Box{
+			Height: ec.ConsoleSize[0],
+			Width:  ec.ConsoleSize[1],
+		}
+	}
```
🤔 I'm thinking now; should the size be part of the `POST /exec/{name:.*}/start` endpoint? Or should it be on both `POST /containers/{name:.*}/exec` and  `POST /exec/{name:.*}/start` ?

Basically thinking (and I think we were discussing that at some point);

- there could be a time between _creating_ and _starting_ the exec
- possibly depending on the scenario, it may be important to have the correct size when _starting_ the exec.
- In some scenarios (possibly) no new size may be passed to that request, in which case we can use the size that's already set during `exec create` 👈 this is what's currently implemented in this PR
- In other scenarios, the `exec start` request _may_ pass a new size. I think we could facilitate that by also adding the size to the `exec start` endpoint; this will be "optional"; if no new size is passed, use what was already set, otherwise update the size when starting the `exec`

vvoland replied:
Yeah I think it makes sense to also be a part of the `start` endpoint. Also the same could apply to the `container start`.

thaJeztah replied:
Yup, looking at it, I think It's probably safe to set it for both `create` and `start` (and both for `container` and `exec`)

vvoland replied:
Done

thaJeztah reviewed (changes requested):
Thanks! Sorry for the delay; I recall I had some thoughts when looking at the changes earlier, but needed to take a bit of time to try to phrase them 😅 

Left my comments; happy to hear your (and other's) thoughts on them 👍

vvoland committed:
container/exec: Support ConsoleSize

Now client have the possibility to set the console size of the executed
process immediately at the creation. This makes a difference for example
when executing commands that output some kind of text user interface
which is bounded by the console dimensions.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on api/types/client.go:
```diff
@@ -57,6 +57,14 @@ type ContainerExecInspect struct {
 	Pid         int
 }
 
+// ContainerExecStartOptions holds the options to start container's exec
+type ContainerExecStartOptions struct {
```
I realize we already have a bunch of these in the top `api/types` package, but I've been trying to move more types into sub packages that are specific to what it's related to, to allow some code to only import things related to that, e.g. `container.StopOptions`; https://github.com/moby/moby/blob/74286cba8c2c14d5360caec44d2da74858e9c707/api/types/container/config.go#L16-L17

Given that this is a new type, perhaps we should move this into `api/types/container` (and name it `container.ExecStartOptions`)

thaJeztah reviewed (changes requested):
one suggestion, but otherwise looks good

I was writing along while reviewing, so let me push my changes as a PR against your branch

thaJeztah commented:
Oh! Didn't comment here; I opened https://github.com/vvoland/moby/pull/1 with my suggestion

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
@ndeloof @rumpl ptal

corhere commented on api/swagger.yaml:
```diff
@@ -9247,6 +9255,7 @@ paths:
               AttachStderr: true
               DetachKeys: "ctrl-p,ctrl-q"
               Tty: false
+              ConsoleSize: nil
```
I'm pretty sure that `nil` is not an alias for null in YAML, so this construct would be equivalent to the JSON `{"ConsoleSize": "nil"}`. I don't think that's your intention. Maybe omit it entirely from the `Tty: false` example and provide a separate example with `Tty: true` and `ConsoleSize` set to an array, like the ExecStartConfig example below?

vvoland replied:
Thanks, good catch! See what Go does to some people... 😄 
Unfortunately multiple examples seems to be supported in OpenAPI 3.0 and we use 2.0 here (https://swagger.io/docs/specification/adding-examples/).
So I'll just leave the `Tty: false` and remove ConsoleSize from here. So we don't repeat the same example as ExecStartConfig.

thaJeztah reviewed (approved):
all green now

still LGTM
